packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD1.hpp

```diff
@@ -0,0 +1,985 @@
+      mapWithCopies = rcp(new map_t(dummy, 
+                                Teuchos::arrayViewFromVector(ownedPlusGhosts),
+                                0, comm)); 
+                                      
+      //create the FEMultiVector for the distributed communication.
+      //We also use the views from this datastructure as arguments to
+      //KokkosKernels coloring functions.
+      if(verbose)std::cout<<comm->getRank()<<": creating FEMultiVector\n";
+      typedef Tpetra::Import<lno_t, gno_t> import_t;
+      Teuchos::RCP<import_t> importer = rcp(new import_t(mapOwned, 
+                                                            mapWithCopies));
+      Teuchos::RCP<femv_t> femv = rcp(new femv_t(mapOwned, 
+                                                    importer, 1, true));
+      //Get color array to fill
+      ArrayRCP<int> colors = solution->getColorsRCP();
+      for(size_t i=0; i<nVtx; i++){
+        colors[i] = 0;
+      } 
+      
+      //Create random numbers seeded on global IDs so that we don't
+      //need to communicate for consistency. These numbers determine
+      //which vertex gets recolored in the event of a conflict.
+      //taken directly from the Zoltan coloring implementation 
+      std::vector<int> rand(finalGIDs.size());
+      for(size_t i = 0; i < finalGIDs.size(); i++){
+        std::srand(finalGIDs[i]);
+        rand[i] = std::rand();
+      }
+
+      std::vector<int> ghostOwners(finalGIDs.size() - nVtx);
+      std::vector<gno_t> ghosts(finalGIDs.size() - nVtx);
+      for(size_t i = nVtx; i < finalGIDs.size(); i++) ghosts[i-nVtx] = finalGIDs[i];
+      ArrayView<int> owners = Teuchos::arrayViewFromVector(ghostOwners);
+      ArrayView<const gno_t> ghostGIDs = Teuchos::arrayViewFromVector(ghosts);
+      mapOwned->getRemoteIndexList(ghostGIDs, owners);
+      
+      for(size_t i = 0; i < finalOffset_vec.size()-1; i++){
+        std::sort(finalAdjs_vec.begin()+finalOffset_vec[i],finalAdjs_vec.begin()+finalOffset_vec[i+1]);
+      }
+      
+      ArrayView<const offset_t> finalOffsets = Teuchos::arrayViewFromVector(finalOffset_vec);
+      ArrayView<const lno_t> finalAdjs = Teuchos::arrayViewFromVector(finalAdjs_vec);
+      
+      Teuchos::ArrayView<const lno_t> exportLIDs = importer->getExportLIDs();
+      Teuchos::ArrayView<const int> exportPIDs = importer->getExportPIDs();
+      
+      for(int i = 0; i < exportLIDs.size(); i++){
+        procs_to_send[exportLIDs[i]].push_back(exportPIDs[i]);  
+      }
+
+      // call coloring function
+      hybridGMB(nVtx, nInterior, finalAdjs, finalOffsets,colors,femv,finalGIDs,rand,owners,mapWithCopies);
+      
+      //copy colors to the output array.
+      for(int i = 0; i < colors.size(); i++){
+        colors[reorderToLocal[i]] = femv->getData(0)[i];
+      }
+    
+    }
+     
+    void hybridGMB(const size_t nVtx,lno_t nInterior, Teuchos::ArrayView<const lno_t> adjs, 
```

Comment:
A comment would be nice here. Looks like this is the main coloring function. Is it reused anywhere?